src/internal/coredb/projects.go

```diff
@@ -0,0 +1,176 @@
+// Functions that take a querier support both running inside a transaction and outside it.
+type QueryExecer interface {
+	sqlx.QueryerContext
+	sqlx.ExecerContext
+}
+
+// CreateProject creates an entry in the core.projects table.
+func CreateProject(ctx context.Context, queryExecer QueryExecer, project *pfs.ProjectInfo) error {
+	_, err := queryExecer.ExecContext(ctx, "INSERT INTO core.projects (name, description) VALUES ($1, $2);", project.Project.Name, project.Description)
+	//todo: insert project.authInfo into auth table.
+	return errors.Wrap(err, "failed to create project")
+}
+
+// DeleteProject deletes an entry in the core.projects table.
+func DeleteProject(ctx context.Context, queryExecer QueryExecer, projectName string) error {
+	result, err := queryExecer.ExecContext(ctx, "DELETE FROM core.projects WHERE name = $1;", projectName)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete project")
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return errors.Wrap(err, "could not get affected rows")
+	}
+	if rowsAffected == 0 {
+		return errors.Wrap(fmt.Errorf("project %s does not exist", projectName), "failed to delete project")
+	}
+	return nil
+}
+
+func DeleteAllProjects(ctx context.Context, queryExecer QueryExecer) error {
+	_, err := queryExecer.ExecContext(ctx, "TRUNCATE core.projects;")
+	return errors.Wrap(err, "could not delete all project rows")
+}
+
+// GetProject retrieves an entry from the core.projects table by project name.
+func GetProject(ctx context.Context, queryExecer QueryExecer, projectName string) (*pfs.ProjectInfo, error) {
+	return getProject(ctx, queryExecer, "name", projectName)
+}
+
+// GetProjectByID is like GetProject, but retrieves an entry using the row id.
+func GetProjectByID(ctx context.Context, queryExecer QueryExecer, id int64) (*pfs.ProjectInfo, error) {
+	return getProject(ctx, queryExecer, "id", id)
+}
+
+func getProject(ctx context.Context, queryExecer QueryExecer, where string, whereVal interface{}) (*pfs.ProjectInfo, error) {
+	row := queryExecer.QueryRowxContext(ctx, fmt.Sprintf("SELECT name, description, created_at FROM core.projects WHERE %s = $1", where), whereVal)
+	project := &pfs.ProjectInfo{Project: &pfs.Project{}}
+	var createdAt time.Time
+	var err error
+	if err = row.Scan(&project.Project.Name, &project.Description, &createdAt); err != nil {
+		return nil, errors.Wrap(err, "failed scanning project row")
+	}
+	project.CreatedAt, err = types.TimestampProto(createdAt)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed converting project proto timestamp")
+	}
+	return project, nil
+}
+
+// UpdateProject updates all fields of an existing project entry in the core.projects table by name. If 'upsert' is set to true, UpdateProject()
+// will attempt to call CreateProject() if the entry does not exist.
+func UpdateProject(ctx context.Context, queryExecer QueryExecer, project *pfs.ProjectInfo, upsert bool) error {
```

Comment:
I think we want to be careful about `upsert bool` becoming a pattern.  Random booleans hanging off the end of functions to control the behavior of the function is pretty confusing to the reader of this code.  It is clearer to call this "InsertOrUpdateProject".

Reply:
An alternative would be to just have an `Upsert` function and keep the complexity of additional parameters for the unexported function.

Reply:
I'm going to have UpsertProject which will determine the target project based on project name. Update project will require an id which should imply that the project already exists.